misc/metrics/src/bandwidth.rs

```diff
@@ -0,0 +1,312 @@
+    type Dial = MapOk<T::Dial, Box<dyn FnOnce((PeerId, M)) -> (PeerId, Muxer<M>) + Send>>;
+
+    fn listen_on(
+        &mut self,
+        id: ListenerId,
+        addr: Multiaddr,
+    ) -> Result<(), TransportError<Self::Error>> {
+        self.transport.listen_on(id, addr)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        self.transport.remove_listener(id)
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let metrics = ConnectionMetrics::from_family_and_addr(&self.metrics, &addr);
+        Ok(self
+            .transport
+            .dial(addr.clone())?
+            .map_ok(Box::new(|(peer_id, stream_muxer)| {
+                (peer_id, Muxer::new(stream_muxer, metrics))
+            })))
+    }
+
+    fn dial_as_listener(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let metrics = ConnectionMetrics::from_family_and_addr(&self.metrics, &addr);
+        Ok(self
+            .transport
+            .dial_as_listener(addr.clone())?
+            .map_ok(Box::new(|(peer_id, stream_muxer)| {
+                (peer_id, Muxer::new(stream_muxer, metrics))
+            })))
+    }
+
+    fn address_translation(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        self.transport.address_translation(server, observed)
+    }
+
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>> {
+        let this = self.project();
+        match this.transport.poll(cx) {
+            Poll::Ready(TransportEvent::Incoming {
+                listener_id,
+                upgrade,
+                local_addr,
+                send_back_addr,
+            }) => {
+                let metrics =
+                    ConnectionMetrics::from_family_and_addr(this.metrics, &send_back_addr);
+                Poll::Ready(TransportEvent::Incoming {
+                    listener_id,
+                    upgrade: upgrade.map_ok(Box::new(|(peer_id, stream_muxer)| {
+                        (peer_id, Muxer::new(stream_muxer, metrics))
+                    })),
+                    local_addr,
+                    send_back_addr,
+                })
+            }
+            Poll::Ready(other) => {
+                let mapped = other.map_upgrade(|_upgrade| unreachable!("case already matched"));
+                Poll::Ready(mapped)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ConnectionMetrics {
+    outbound: Counter,
+    inbound: Counter,
+}
+
+impl ConnectionMetrics {
+    fn from_family_and_addr(family: &Family<Labels, Counter>, protocols: &Multiaddr) -> Self {
+        let protocols = protocol_stack::as_string(protocols);
+
+        // Additional scope to make sure to drop the lock guard from `get_or_create`.
```

Comment:
Is this still needed if you just chain the `.clone()` directly onto it?

Reply:
The mutex guard would still be in scope and thus an additional `lock` in the instruction right after would deadlock.